flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java

```diff
@@ -299,6 +303,8 @@ public void start(ResourceManagerId newResourceManagerId, Executor newMainThread
 			TimeUnit.MILLISECONDS);
 
 		registerSlotManagerMetrics();
+
+		allocateRedundantSlots(redundantSlotNum);
```

Comment:
After offline discussion, SlotManagerImpl will not allocate redundant taskmanagers at start for that it is hard to know wether there exist enough redundant taskmanagers.

Reply:
I'd like to explain a bit more about the outcome of the offline discussion, for public visibility.

We do not allocate redundant task managers immediately when starting the resource manager, to leave some time for task managers from previous attempts to register. Otherwise, we may result in holding more task managers than expected (required + redundant), and those extra task managers may not be easily released if there are allocated slots.

Instead of allocating redundant task managers immediately, we periodically check number of task managers, together with checking task manager idleness, and allocate redundant task managers if needed.